# Problem 1

### Import Data: CMT Rates

In [1]:
import numpy as np
import pandas as pd
import quandl

In [2]:
token = 'PxHgKF5Bzq8ypAUmtKsg'
CMTs = quandl.get("USTREASURY/YIELD", authtoken=token, start_date="2013-01-01", end_date="2017-01-31")

In [3]:
CMTs.head()

1 MO  2 MO  3 MO  6 MO  1 YR  2 YR  3 YR  5 YR  7 YR  10 YR  \
Date                                                                      
2013-01-02  0.07  None  0.08  0.12  0.15  0.27  0.37  0.76  1.25   1.86   
2013-01-03  0.06  None  0.08  0.12  0.15  0.27  0.40  0.81  1.31   1.92   
2013-01-04  0.06  None  0.07  0.11  0.15  0.27  0.41  0.82  1.32   1.93   
2013-01-07  0.05  None  0.07  0.11  0.15  0.27  0.41  0.82  1.31   1.92   
2013-01-08  0.06  None  0.07  0.11  0.14  0.25  0.38  0.79  1.28   1.89   

            20 YR  30 YR  
Date                      
2013-01-02   2.63   3.04  
2013-01-03   2.70   3.12  
2013-01-04   2.70   3.10  
2013-01-07   2.70   3.10  
2013-01-08   2.66   3.06

In [4]:
global CMTs

In [5]:
# Remove '2 MO'
CMTs = CMTs.drop(columns = '2 MO')

## Step 1
* As shown in Diebold-Li paper, we assume that $\lambda = 0.0609 $.

In [6]:
lamb0 = 0.0609 # The lambda assumed
m = [1,3,6,12,12*2,12*3,12*5,12*7,12*10,12*20,12*30]

In [7]:
import math
from statsmodels.regression.linear_model import OLS

In [8]:
exog0 = np.array([[1,(1-math.exp(-lamb0*i))/(lamb0*i),(1-math.exp(-lamb0*i))/(lamb0*i)-math.exp(-lamb0*i)] for i in m])

In [9]:
demo = pd.DataFrame(columns = ['b1','b2','b3','ssr'])
for j in CMTs.index:
    model = OLS(CMTs.loc[j,:], exog0).fit()
    demo.loc[j,:] = np.append(np.array(model.params),model.ssr)

* **Solution to Step 1**:The demo table stores the **fitted $\beta$s** on specific dates and also records the **ssr(Sum Squared of Residuals)**.

In [10]:
demo.head()

b1       b2       b3       ssr
2013-01-02  3.10231 -2.75766 -5.45356  0.455388
2013-01-03  3.18981  -2.8519 -5.53614  0.442416
2013-01-04  3.18349 -2.85776 -5.47936  0.415237
2013-01-07   3.1782 -2.85683 -5.46594  0.427124
2013-01-08  3.13969  -2.8088 -5.48885  0.420724

## Step 2

SSR Function:
* exog: $X$ matrix in linear regression
* b: fitted betas, the formula is $\beta = (X^{T}X)^{-1}X^{T}y$
* e: errors, the formula is $error = y-\tilde{y} = y - X\beta$

In [11]:
def SSR(lam):
    ssr = 0
    m = [1,3,6,12,12*2,12*3,12*5,12*7,12*10,12*20,12*30]
    exog = np.array([[1,(1-math.exp(-lam*i))/(lam*i),(1-math.exp(-lam*i))/(lam*i)-math.exp(-lam*i)] for i in m], dtype='float')
    for j in CMTs.index:
        b = np.dot(np.linalg.inv(exog.T.dot(exog)), exog.T.dot(CMTs.loc[j,:]))
        e = CMTs.loc[j,:]-exog.dot(b)
        ssr = ssr + e.T.dot(e)
    return ssr

In [12]:
import scipy as sp

In [13]:
optimize = sp.optimize.minimize(SSR,lamb0)

In [14]:
optimize

      fun: 57.33477907612194
 hess_inv: array([[2.75080097e-06]])
      jac: array([9.53674316e-07])
  message: 'Optimization terminated successfully.'
     nfev: 33
      nit: 7
     njev: 11
   status: 0
  success: True
        x: array([0.0431258])

* **Solution to Step 2**: It shows that the **optimal $\lambda = 0.0431258$**. In the meanwhile, it's total SSR is about 57.33.

In [15]:
lamb = optimize['x'][0]

In [16]:
exog = np.array([[1,(1-math.exp(-lamb*i))/(lamb*i),(1-math.exp(-lamb*i))/(lamb*i)-math.exp(-lamb*i)] for i in m])

In [17]:
factors = pd.DataFrame(columns = ['b1','b2','b3','ssr'], dtype = 'float')
for j in CMTs.index:
    model = OLS(CMTs.loc[j,:], exog).fit()
    factors.loc[j,:] = np.append(np.array(model.params),model.ssr)

* factors table stores the best fitted dynamic $\beta$s and corresponding ssr.

In [18]:
factors.head()

b1        b2        b3       ssr
2013-01-02  3.423031 -3.200014 -4.891039  0.115209
2013-01-03  3.512073 -3.299002 -4.933246  0.105931
2013-01-04  3.500129 -3.298936 -4.860845  0.095373
2013-01-07  3.495753 -3.297564 -4.855132  0.100726
2013-01-08  3.457841 -3.251722 -4.890358  0.096970

# Problem 2

## Step 1

In [19]:
factors['RMSE'] = np.sqrt(factors['ssr']/len(m))

In [20]:
factors['RMSE'].nlargest(20)

2013-04-04    0.110085
2013-05-02    0.108868
2013-04-03    0.108627
2013-05-01    0.108623
2013-03-27    0.108620
2013-03-28    0.108543
2013-04-26    0.108489
2013-04-30    0.108409
2013-04-29    0.108024
2013-04-05    0.107589
2013-04-22    0.106212
2013-04-09    0.105977
2013-04-17    0.105566
2013-03-01    0.105428
2013-04-24    0.104982
2013-04-18    0.104867
2013-02-28    0.104844
2013-04-15    0.104702
2013-04-23    0.104070
2013-05-16    0.103936
Name: RMSE, dtype: float64

* **Solution to Step 1**: Series above show the largest RMSE and corresponding 20 days and days stores all 20 dates with largest RMSE.

In [21]:
days = factors['RMSE'].nlargest(20).index

## Step 2

According to Diebold-Li's paper, because of two disadvantages of VAR(1), simple **AR(1)** model will be chosen for estimation and forecasts instead of VAR(1).

### Fit AR(1) Model
* AR(1):
$$\beta_{i}(t) = intercept + slope \times \beta_{i}(t-1) + \epsilon_{i}(t) ,i=1,2,3$$
* Forecast using AR(1):
$$\beta_{i}(t) = intercept + slope \times \beta_{i}(t-1) ,i=1,2,3$$

In [22]:
from dateutil.relativedelta import relativedelta

In [23]:
import statsmodels.api as sm
def AR(s):
    """Calculate the AR(1) parameters and Half Life of mean reversion series.
    Methodology: https://quant.stackexchange.com/questions/25086/calculating-half-life-of-mean-reverting-series-with-python
    
    Args:
    s: pandas.Series
    
    Returns:
    Intercept: float
    Slope: float
    HF: float
    """
    x = s.shift()[1:] # Y(t-1)
    y = s[1:] # Y(t)
    x = sm.add_constant(x) #adds intercept terms to X variable for regression
    x.columns = ['const','x'] # Y(t) = intercept + slope *Y(t-1) + e
    model = sm.OLS(y,x).fit()
    
    intercept = model.params[0]
    slope = sm.OLS(y,x).fit().params[1] 
    HF = -math.log(2) / (slope-1) # Half-life = - log(2)/slope
    return intercept, slope, HF

In [47]:
AR_data = pd.DataFrame(columns = np.arange(0,9))
for t in days:
    AR_data.loc[t,np.arange(3)] = AR(factors.loc[t-relativedelta(months = 6):t,'b1'])
    AR_data.loc[t,np.arange(3,6)] = AR(factors.loc[t-relativedelta(months = 6):t,'b2'])
    AR_data.loc[t,np.arange(6,9)] = AR(factors.loc[t-relativedelta(months = 6):t,'b3'])
    
col = []
for i in factors.columns[0:3]:
    for j in ['intercept','slope','HF']:
        col.append(i+'_'+j)
AR_data.columns = col

* AR_data stores AR(1) parameters of all $\beta$s on the 20 dates.

In [25]:
AR_data

b1_intercept  b1_slope    b1_HF b2_intercept  b2_slope    b2_HF  \
2013-04-04     0.368666  0.896024  6.66639    -0.408237  0.877282   5.6483   
2013-05-02    0.0726951  0.978275  31.9052    -0.111103  0.965231  19.9355   
2013-04-03     0.434377  0.877897  5.67673    -0.478384  0.856701  4.83706   
2013-05-01    0.0803295  0.976127  29.0344    -0.115176   0.96401  19.2595   
2013-03-27     0.463961  0.869968   5.3306    -0.497596  0.851316  4.66188   
2013-03-28      0.46094  0.870787  5.36437    -0.503638  0.849564  4.60758   
2013-04-26     0.111502  0.967394   21.258    -0.148755  0.954001  15.0688   
2013-04-30     0.114161  0.966645  20.7812    -0.150751  0.953401  14.8749   
2013-04-29     0.119269  0.965214  19.9261    -0.157014  0.951534  14.3017   
2013-04-05     0.176168  0.949549   13.739    -0.239677  0.927215  9.52321   
2013-04-22     0.135983  0.960566  17.5772    -0.179564  0.944872  12.5734   
2013-04-09     0.275944  0.921657   8.8476    -0.312227  0.905618  7.34408   
2013-04-17     0.175533  0.949492  13.7235    -0.215541   0.93415  10.5262   
2013-03-01     0.464104   0.86907  5.29405    -0.477598  0.856398  4.82688   
2013-04-24     0.135773  0.960616  17.5996     -0.16901  0.948002  13.3303   
2013-04-18     0.140046  0.959428  17.0845    -0.194615  0.940386  11.6274   
2013-02-28     0.480269  0.864843  5.12847    -0.510838  0.846781  4.52389   
2013-04-15     0.192271  0.944826   12.563    -0.239002  0.927176  9.51816   
2013-04-23     0.151332  0.956265  15.8488    -0.190973  0.941471  11.8427   
2013-05-16     0.205359  0.941139   11.776    -0.233141  0.929045  9.76883   

           b3_intercept  b3_slope    b3_HF  
2013-04-04    -0.414536  0.917937  8.44656  
2013-05-02    -0.257953   0.94758  13.2231  
2013-04-03    -0.411582  0.918576  8.51285  
2013-05-01    -0.282674  0.942709  12.0988  
2013-03-27    -0.383588  0.924302  9.15679  
2013-03-28    -0.408222  0.919226  8.58132  
2013-04-26    -0.289675  0.941348   11.818  
2013-04-30    -0.293758  0.940527  11.6549  
2013-04-29    -0.310737  0.937194  11.0363  
2013-04-05    -0.441922  0.911545  7.83615  
2013-04-22    -0.277601  0.943748  12.3221  
2013-04-09    -0.535734  0.893071   6.4823  
2013-04-17    -0.372291  0.925029  9.24559  
2013-03-01    -0.449703  0.910281  7.72576  
2013-04-24    -0.304653  0.938393   11.251  
2013-04-18    -0.257001   0.94783  13.2864  
2013-02-28    -0.406853  0.919053  8.56293  
2013-04-15     -0.47166  0.905451  7.33112  
2013-04-23    -0.306778  0.937973  11.1749  
2013-05-16    -0.335481  0.932308  10.2397

### Forecast $\beta$

In [26]:
def forecast_AR(date,data_set,AR):
    forecast = pd.DataFrame(columns = ['b'+str(i) for i in range(1,4)])
    for t in data_set.loc[i+relativedelta(days=1):i+relativedelta(months=1,days=1),:].index:
        for j in forecast.columns:
            forecast.loc[t,j] = AR.loc[i,j+'_intercept']+AR.loc[i,j+'_slope']*data_set.shift().loc[t,j]
    return forecast

In [27]:
forecast_b = {}
for i in days:
    forecast_b[i] = forecast_AR(i, factors, AR_data)

* **Solution to Step 2**: forecast_b stores t+1m forecasted $\beta$s on the 20 dates.

In [28]:
forecast_b[days[0]] # Demo: Forecast betas at 2013-04-04

b1       b2       b3
2013-04-05  3.39508   -3.169 -5.04231
2013-04-08  3.26795 -3.06174 -4.76566
2013-04-09  3.31331 -3.10883 -4.79693
2013-04-10  3.34841 -3.13908 -4.87543
2013-04-11  3.42427 -3.21106 -4.95383
2013-04-12  3.41553 -3.20444 -4.95158
2013-04-15  3.32324 -3.12244 -4.85964
2013-04-16  3.27714 -3.08053 -4.78651
2013-04-17  3.30909 -3.10587  -4.8006
2013-04-18  3.27959  -3.0828 -4.72829
2013-04-19  3.25616 -3.07426 -4.65614
2013-04-22  3.26909 -3.08405 -4.67406
2013-04-23  3.26765  -3.0793   -4.701
2013-04-24  3.29303 -3.10391 -4.74257
2013-04-25  3.27936 -3.08096 -4.75322
2013-04-26  3.29894 -3.11216 -4.74717
2013-04-29  3.25314 -3.05915 -4.75803
2013-04-30  3.26973 -3.07945 -4.79966
2013-05-01   3.2677 -3.07747 -4.77831
2013-05-02  3.21806  -3.0247  -4.7696
2013-05-03  3.21103 -3.02566 -4.73785

### Forecast yields

In [29]:
def DNS2Y(beta_set, lamb, months):
    """
    Convert DNS model betas to CMT yields.
    """
    forecast = pd.DataFrame(columns = months.keys())
    for t in beta_set.index:
        for m,n in months.items():
            forecast.loc[t,m] = beta_set.loc[t,'b1'] + beta_set.loc[t,'b2']*(1-math.exp(-lamb*n))/(lamb*n) + beta_set.loc[t,'b3']*((1-math.exp(-lamb*n))/(lamb*n)-math.exp(-lamb*n))
    return forecast

In [30]:
month = dict(zip(CMTs.columns,m))
forecast_y = {}
for i in days:
    forecast_y[i] = DNS2Y(forecast_b[i], lamb, month)

* **Soluntion to Step 2**: forecast_y stores t+1m forecasted CMT yields on the 20 dates.

In [31]:
forecast_y[days[0]].head(5) # Forecast t+5d CMTs Rates from 2013-04-04 

1 MO      3 MO       6 MO         1 YR       2 YR      3 YR  \
2013-04-05  0.187797  0.123157  0.0528633  -0.00989044  0.0708576  0.293356   
2013-04-08  0.171432  0.113052  0.0504056  -0.00221493  0.0846067  0.302572   
2013-04-09  0.170058  0.112394  0.0508744  0.000620796    0.09227  0.314777   
2013-04-10  0.173907  0.114459  0.0507585  -0.00238953  0.0875064  0.311277   
2013-04-11  0.177673  0.118145  0.0546438   0.00280033  0.0975321  0.327379   

                5 YR     7 YR    10 YR    20 YR    30 YR  
2013-04-05  0.839548  1.32363  1.84588  2.60192  2.86619  
2013-04-08  0.828812  1.29225  1.79094  2.51187  2.76378  
2013-04-09  0.848528   1.3174  1.82143  2.54967   2.8041  
2013-04-10  0.850662  1.32537  1.83608  2.57426  2.83219  
2013-04-11   0.87867  1.36292  1.88348  2.63559  2.89836

In [32]:
forecast_y[days[0]].head(10) # Forecast t+10d CMTs Rates from 2013-04-04 

1 MO      3 MO       6 MO         1 YR       2 YR      3 YR  \
2013-04-05  0.187797  0.123157  0.0528633  -0.00989044  0.0708576  0.293356   
2013-04-08  0.171432  0.113052  0.0504056  -0.00221493  0.0846067  0.302572   
2013-04-09  0.170058  0.112394  0.0508744  0.000620796    0.09227  0.314777   
2013-04-10  0.173907  0.114459  0.0507585  -0.00238953  0.0875064  0.311277   
2013-04-11  0.177673  0.118145  0.0546438   0.00280033  0.0975321  0.327379   
2013-04-12  0.175451   0.11574  0.0519743  -0.00036311   0.093513  0.322662   
2013-04-15   0.16535  0.105831  0.0419682   -0.0116578  0.0769266  0.299228   
2013-04-16  0.161807  0.103391  0.0407776   -0.0115321  0.0765189  0.296059   
2013-04-17  0.168644  0.110719  0.0488484  -0.00198007  0.0888974   0.31096   
2013-04-18  0.163242  0.107155  0.0475657 -0.000115402  0.0931324  0.314582   

                5 YR     7 YR    10 YR    20 YR    30 YR  
2013-04-05  0.839548  1.32363  1.84588  2.60192  2.86619  
2013-04-08  0.828812  1.29225  1.79094  2.51187  2.76378  
2013-04-09  0.848528   1.3174  1.82143  2.54967   2.8041  
2013-04-10  0.850662  1.32537  1.83608  2.57426  2.83219  
2013-04-11   0.87867  1.36292  1.88348  2.63559  2.89836  
2013-04-12  0.872927   1.3565  1.87643  2.62771  2.89019  
2013-04-15  0.835893  1.30849  1.81705  2.55222  2.80911  
2013-04-16  0.825411  1.29135  1.79264  2.51723  2.77042  
2013-04-17  0.844318  1.31307  1.81708  2.54537  2.79982  
2013-04-18  0.843471  1.30727   1.8055  2.52508  2.77647

In [33]:
forecast_y[days[0]] # Forecast t+1m CMTs Rates from 2013-04-04 

1 MO       3 MO       6 MO         1 YR       2 YR      3 YR  \
2013-04-05  0.187797   0.123157  0.0528633  -0.00989044  0.0708576  0.293356   
2013-04-08  0.171432   0.113052  0.0504056  -0.00221493  0.0846067  0.302572   
2013-04-09  0.170058   0.112394  0.0508744  0.000620796    0.09227  0.314777   
2013-04-10  0.173907   0.114459  0.0507585  -0.00238953  0.0875064  0.311277   
2013-04-11  0.177673   0.118145  0.0546438   0.00280033  0.0975321  0.327379   
2013-04-12  0.175451    0.11574  0.0519743  -0.00036311   0.093513  0.322662   
2013-04-15   0.16535   0.105831  0.0419682   -0.0116578  0.0769266  0.299228   
2013-04-16  0.161807   0.103391  0.0407776   -0.0115321  0.0765189  0.296059   
2013-04-17  0.168644   0.110719  0.0488484  -0.00198007  0.0888974   0.31096   
2013-04-18  0.163242   0.107155  0.0475657 -0.000115402  0.0931324  0.314582   
2013-04-19  0.149687  0.0960245  0.0395357  -0.00355109  0.0943446   0.31685   
2013-04-22  0.152666  0.0987136  0.0418909  -0.00156707  0.0963831   0.31951   
2013-04-23  0.155308   0.100127  0.0416953  -0.00427524  0.0906879  0.312506   
2013-04-24  0.155727  0.0999516  0.0408534  -0.00578726  0.0896018  0.313079   
2013-04-25  0.164294   0.107175  0.0462415  -0.00350849  0.0873774  0.307909   
2013-04-26  0.153463  0.0978475  0.0389909  -0.00716778  0.0891414   0.31344   
2013-04-29  0.159314   0.101122  0.0387078   -0.0135959  0.0734517  0.291335   
2013-04-30  0.155167  0.0962018  0.0328738   -0.0205373  0.0662518  0.285296   
2013-05-01  0.155523   0.097297  0.0349174   -0.0170796  0.0711709  0.290591   
2013-05-02  0.157716  0.0976762  0.0327264   -0.0239119  0.0567418  0.270329   
2013-05-03  0.150407  0.0916257  0.0283092    -0.025833  0.0576106  0.272206   

                5 YR     7 YR    10 YR    20 YR    30 YR  
2013-04-05  0.839548  1.32363  1.84588  2.60192  2.86619  
2013-04-08  0.828812  1.29225  1.79094  2.51187  2.76378  
2013-04-09  0.848528   1.3174  1.82143  2.54967   2.8041  
2013-04-10  0.850662  1.32537  1.83608  2.57426  2.83219  
2013-04-11   0.87867  1.36292  1.88348  2.63559  2.89836  
2013-04-12  0.872927   1.3565  1.87643  2.62771  2.89019  
2013-04-15  0.835893  1.30849  1.81705  2.55222  2.80911  
2013-04-16  0.825411  1.29135  1.79264  2.51723  2.77042  
2013-04-17  0.844318  1.31307  1.81708  2.54537  2.79982  
2013-04-18  0.843471  1.30727   1.8055  2.52508  2.77647  
2013-04-19  0.843456  1.30359  1.79717  2.50944  2.75823  
2013-04-22  0.847836  1.30956  1.80488   2.5197  2.76939  
2013-04-23   0.84049  1.30288  1.79933  2.51612  2.76652  
2013-04-24  0.845338  1.31158  1.81222   2.5351  2.78763  
2013-04-25  0.836866   1.3015  1.80098  2.52262  2.77475  
2013-04-26  0.847004  1.31417  1.81569  2.53977  2.79271  
2013-04-29   0.81708  1.27998  1.77805  2.49804  2.74962  
2013-04-30  0.814668  1.28104  1.78298  2.50865  2.76223  
2013-05-01   0.81937  1.28471  1.78531  2.50887   2.7617  
2013-05-02  0.791054  1.25136  1.74744  2.46518  2.71603  
2013-05-03  0.792637  1.25175  1.74614  2.46112  2.71097

## Step 3

### Forecast $\beta$ with Random Walk
* Random Walk:
$$\beta_{i}(t) = \beta_{i}(t-1) + \epsilon_{i}(t) ,i=1,2,3$$
* Forecast using AR(1):
$$\beta_{i}(t) = \beta_{i}(t-1) ,i=1,2,3$$

In [34]:
def forecast_RW(date,date_set):
    forecast = pd.DataFrame(columns = ['b'+str(i) for i in range(1,4)])
    for t in factors.loc[i+relativedelta(days=1):i+relativedelta(months=1,days=1),:].index:
        for j in forecast.columns:
            forecast.loc[t,j] = factors.shift().loc[t,j]
    return forecast

* forecast_b stores t+1m forecasted $\beta$s on the 20 dates using Random Walk Model.

In [35]:
forecast_b2 = {}
for i in days:
    forecast_b2[i] = forecast_RW(i, factors)

### Forecast yields with Random Walk $\beta$

In [36]:
forecast_y2 = {}
for i in days:
    forecast_y2[i] = DNS2Y(forecast_b2[i], lamb, month)

* forecast_y stores t+1m forecasted CMT yields on the 20 dates using Random Walk Model.

In [37]:
forecast_y2[days[0]] # Forecast t+1m CMTs Rates from 2013-04-04

1 MO       3 MO       6 MO         1 YR       2 YR      3 YR  \
2013-04-05  0.191921   0.126414  0.0549061   -0.0100004  0.0673386  0.287319   
2013-04-08  0.176005   0.117097  0.0536124  -0.00079894  0.0823027  0.296722   
2013-04-09   0.17339    0.11536  0.0532373   0.00164325  0.0903789  0.310019   
2013-04-10  0.177024   0.117112  0.0527007  -0.00189315  0.0851733  0.306367   
2013-04-11  0.179591    0.11974  0.0557526   0.00294494  0.0958526  0.324084   
2013-04-12  0.177264   0.117201  0.0529057 -0.000473577  0.0914493  0.318881   
2013-04-15  0.167853    0.10783  0.0431939    -0.012005  0.0735003  0.292996   
2013-04-16  0.164837   0.105928  0.0425331   -0.0114443  0.0731462   0.28939   
2013-04-17  0.171884   0.113563  0.0510489  -0.00118633  0.0866851  0.305824   
2013-04-18  0.166352   0.109986  0.0498919  0.000968389  0.0912385  0.309575   
2013-04-19  0.151384  0.0976411  0.0408999   -0.0030619  0.0922035  0.311641   
2013-04-22   0.15449   0.100452  0.0433749  -0.00094168   0.094461  0.314632   
2013-04-23  0.157564   0.102178  0.0433346  -0.00374322  0.0883676  0.307085   
2013-04-24  0.157481   0.101498  0.0419989  -0.00568417  0.0870865  0.307755   
2013-04-25  0.167584   0.110089  0.0485252   -0.0026615  0.0850208  0.302347   
2013-04-26  0.154765  0.0989719  0.0397601  -0.00735595  0.0864829  0.308094   
2013-04-29  0.162537   0.103829  0.0405893     -0.01349  0.0698373  0.284152   
2013-04-30  0.157458  0.0979499  0.0337729   -0.0214098  0.0617975  0.277495   
2013-05-01   0.15789  0.0991827  0.0360332    -0.017619  0.0671645  0.283259   
2013-05-02  0.161564   0.100772  0.0346712   -0.0243049  0.0517826  0.261216   
2013-05-03  0.153367  0.0939479  0.0296289    -0.026623  0.0525115  0.263049   

                5 YR     7 YR    10 YR    20 YR    30 YR  
2013-04-05  0.830185  1.31228   1.8328  2.58666  2.85019  
2013-04-08  0.817038  1.27615   1.7706  2.48569  2.73558  
2013-04-09  0.838875  1.30413  1.80456  2.52784  2.78056  
2013-04-10  0.841589  1.31334  1.82116  2.55542    2.812  
2013-04-11  0.872824  1.35529  1.87413  2.62391  2.88588  
2013-04-12  0.866454  1.34815  1.86628  2.61512  2.87677  
2013-04-15   0.82513  1.29451  1.79992  2.53083  2.78624  
2013-04-16  0.813239  1.27517  1.77251  2.49169    2.743  
2013-04-17  0.834229  1.29934  1.79975  2.52306   2.7758  
2013-04-18  0.832937  1.29255  1.78657  2.50029  2.74965  
2013-04-19  0.832451  1.28803  1.77695  2.48267  2.72919  
2013-04-22   0.83739  1.29475   1.7856  2.49415  2.74165  
2013-04-23  0.829435  1.28749  1.77956  2.49023   2.7385  
2013-04-24  0.834948  1.29732  1.79404  2.51146   2.7621  
2013-04-25  0.825769  1.28628  1.78165  2.49761  2.74778  
2013-04-26  0.836771  1.30018   1.7979  2.51667  2.76777  
2013-04-29  0.803909  1.26243  1.75618  2.47025  2.71978  
2013-04-30  0.801357  1.26376   1.7618  2.48214  2.73388  
2013-05-01  0.806461  1.26772   1.7643  2.48234  2.73326  
2013-05-02  0.775244  1.23079  1.72224  2.43368  2.68235  
2013-05-03  0.776763  1.23101  1.72063  2.42906  2.67666

### Evaluation

In [38]:
def forecast_performance(forecast_set, data_set = CMTs):
    forecast_error = data_set.loc[forecast_set.index,:] - forecast_set
    forecast_perform = pd.DataFrame(columns = ['mean','std','RMSE'])
    forecast_perform.loc[:,'mean'] = forecast_error.mean()
    forecast_perform.loc[:,'std'] = forecast_error.std()
    forecast_perform.loc[:,'RMSE'] = np.sqrt((forecast_error**2).mean())
    
    return forecast_perform

#### Out-of-sample forecast performance of AR(1)

In [39]:
# t+5d forceast performance on 2013-04-04
forecast_performance(forecast_y[days[0]].head(5), CMTs)

mean       std      RMSE
1 MO  -0.118173  0.012897  0.118735
3 MO  -0.048242  0.004546  0.048413
6 MO   0.048091  0.001806  0.048118
1 YR   0.128215  0.008788  0.128455
2 YR   0.153445  0.010053  0.153709
3 YR   0.034128  0.010511  0.035399
5 YR  -0.135244  0.020329  0.136461
7 YR  -0.156314  0.032240  0.158951
10 YR -0.051563  0.048145  0.067179
20 YR -0.002661  0.061049  0.054669
30 YR  0.115077  0.066578  0.129572

In [40]:
# t+10d forceast performance on 2013-04-04
forecast_performance(forecast_y[days[0]].head(10), CMTs)

mean       std      RMSE
1 MO  -0.119536  0.011768  0.120056
3 MO  -0.049404  0.005167  0.049647
6 MO   0.045932  0.004071  0.046094
1 YR   0.127672  0.009965  0.128022
2 YR   0.149824  0.011818  0.150243
3 YR   0.030715  0.011772  0.032682
5 YR  -0.137824  0.019769  0.139094
7 YR  -0.168826  0.030965  0.171362
10 YR -0.069652  0.042384  0.080425
20 YR -0.021092  0.051163  0.052922
30 YR  0.099937  0.054105  0.112348

In [41]:
# t+1m forceast performance on 2013-04-04
forecast_performance(forecast_y[days[0]], CMTs)

mean       std      RMSE
1 MO  -0.120221  0.010632  0.120667
3 MO  -0.047991  0.005414  0.048281
6 MO   0.046789  0.005155  0.047059
1 YR   0.128264  0.007697  0.128483
2 YR   0.145499  0.011683  0.145945
3 YR   0.031148  0.013915  0.033979
5 YR  -0.136383  0.024012  0.138381
7 YR  -0.170018  0.032611  0.172971
10 YR -0.071749  0.040712  0.082015
20 YR -0.016455  0.049419  0.050958
30 YR  0.111886  0.050656  0.122320

* From above, we can see the more time interval we forecast, generally the more RMSE it will be.

#### Out-of-sample forecast performance of Random Walk

In [42]:
# t+5d forceast performance on 2013-04-04
forecast_performance(forecast_y2[days[0]].head(5), CMTs)

mean       std      RMSE
1 MO  -0.121586  0.013381  0.122174
3 MO  -0.051145  0.004476  0.051301
6 MO   0.045958  0.001255  0.045972
1 YR   0.127621  0.008780  0.127862
2 YR   0.155791  0.010757  0.156088
3 YR   0.039098  0.011163  0.040353
5 YR  -0.126102  0.020932  0.127485
7 YR  -0.144238  0.033278  0.147277
10 YR -0.036650  0.050018  0.057834
20 YR  0.016096  0.064262  0.059689
30 YR  0.135157  0.069831  0.148891

In [43]:
# t+10d forceast performance on 2013-04-04
forecast_performance(forecast_y2[days[0]].head(10), CMTs)

mean       std      RMSE
1 MO  -0.122612  0.011907  0.123131
3 MO  -0.052023  0.005070  0.052245
6 MO   0.044022  0.003854  0.044173
1 YR   0.127225  0.009987  0.127577
2 YR   0.152293  0.012233  0.152735
3 YR   0.035882  0.012265  0.037722
5 YR  -0.128250  0.020956  0.129782
7 YR  -0.156090  0.032241  0.159059
10 YR -0.053828  0.044046  0.068143
20 YR -0.001051  0.053893  0.051138
30 YR  0.121432  0.057211  0.133010

In [44]:
# t+1m forceast performance on 2013-04-04 model:Random Walk
forecast_performance(forecast_y2[days[0]], CMTs)

mean       std      RMSE
1 MO  -0.122962  0.011024  0.123432
3 MO  -0.050321  0.005607  0.050618
6 MO   0.045125  0.005347  0.045426
1 YR   0.128054  0.007874  0.128285
2 YR   0.148391  0.011940  0.148848
3 YR   0.037052  0.014514  0.039667
5 YR  -0.125381  0.025215  0.127773
7 YR  -0.155270  0.033807  0.158736
10 YR -0.053292  0.042155  0.067323
20 YR  0.007106  0.052053  0.051293
30 YR  0.137210  0.053973  0.146973

* Compare two forecast methods on 2013-04-04, it is easy to find that **the AR(1) model has more accuracy (lower RMSE) than Random Walk model**.

#### Compare two methods on different dates

In [45]:
from IPython.display import display_html
def display_side_by_side(*args):
    # Display DataFrames side by side, origin: https://stackoverflow.com/questions/38783027/jupyter-notebook-display-two-pandas-tables-side-by-side
    html_str=''
    for df in args:
        html_str+=df.to_html()
    display_html(html_str.replace('table','table style="display:inline"'),raw=True)

In [46]:
for i in range(20):
    df1 = forecast_performance(forecast_y[days[i]], CMTs)
    df2 = forecast_performance(forecast_y2[days[i]], CMTs)
    display_side_by_side(df1,df2)

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.120221 
 0.010632 
 0.120667 
 
 
 3 MO 
 -0.047991 
 0.005414 
 0.048281 
 
 
 6 MO 
 0.046789 
 0.005155 
 0.047059 
 
 
 1 YR 
 0.128264 
 0.007697 
 0.128483 
 
 
 2 YR 
 0.145499 
 0.011683 
 0.145945 
 
 
 3 YR 
 0.031148 
 0.013915 
 0.033979 
 
 
 5 YR 
 -0.136383 
 0.024012 
 0.138381 
 
 
 7 YR 
 -0.170018 
 0.032611 
 0.172971 
 
 
 10 YR 
 -0.071749 
 0.040712 
 0.082015 
 
 
 20 YR 
 -0.016455 
 0.049419 
 0.050958 
 
 
 30 YR 
 0.111886 
 0.050656 
 0.122320 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122962 
 0.011024 
 0.123432 
 
 
 3 MO 
 -0.050321 
 0.005607 
 0.050618 
 
 
 6 MO 
 0.045125 
 0.005347 
 0.045426 
 
 
 1 YR 
 0.128054 
 0.007874 
 0.128285 
 
 
 2 YR 
 0.148391 
 0.011940 
 0.148848 
 
 
 3 YR 
 0.037052 
 0.014514 
 0.039667 
 
 
 5 YR 
 -0.125381 
 0.025215 
 0.127773 
 
 
 7 YR 
 -0.155270 
 0.033807 
 0.158736 
 
 
 10 YR 
 -0.053292 
 0.042155 
 0.067323 
 
 
 20 YR 
 0.007106 
 0.052053 
 0.051293 
 
 
 30 YR 
 0.137210 
 0.053973 
 0.146973

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.114240 
 0.017464 
 0.115504 
 
 
 3 MO 
 -0.041120 
 0.007924 
 0.041841 
 
 
 6 MO 
 0.047345 
 0.011272 
 0.048606 
 
 
 1 YR 
 0.117271 
 0.014975 
 0.118179 
 
 
 2 YR 
 0.128630 
 0.025754 
 0.131063 
 
 
 3 YR 
 0.030855 
 0.026185 
 0.040063 
 
 
 5 YR 
 -0.085116 
 0.041313 
 0.094182 
 
 
 7 YR 
 -0.104266 
 0.050256 
 0.115225 
 
 
 10 YR 
 -0.013819 
 0.051869 
 0.052472 
 
 
 20 YR 
 0.027462 
 0.057632 
 0.062590 
 
 
 30 YR 
 0.144336 
 0.063558 
 0.157100 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.112344 
 0.019404 
 0.113928 
 
 
 3 MO 
 -0.039565 
 0.008942 
 0.040516 
 
 
 6 MO 
 0.048432 
 0.010244 
 0.049453 
 
 
 1 YR 
 0.117560 
 0.014012 
 0.118353 
 
 
 2 YR 
 0.127745 
 0.025586 
 0.130162 
 
 
 3 YR 
 0.029185 
 0.026209 
 0.038807 
 
 
 5 YR 
 -0.087696 
 0.040987 
 0.096388 
 
 
 7 YR 
 -0.107312 
 0.049884 
 0.117838 
 
 
 10 YR 
 -0.017220 
 0.052059 
 0.053644 
 
 
 20 YR 
 0.023663 
 0.058924 
 0.062182 
 
 
 30 YR 
 0.140409 
 0.065587 
 0.154309

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.119349 
 0.010550 
 0.119794 
 
 
 3 MO 
 -0.047846 
 0.005232 
 0.048118 
 
 
 6 MO 
 0.046649 
 0.005013 
 0.046905 
 
 
 1 YR 
 0.127681 
 0.007688 
 0.127901 
 
 
 2 YR 
 0.143025 
 0.011601 
 0.143473 
 
 
 3 YR 
 0.027616 
 0.013752 
 0.030711 
 
 
 5 YR 
 -0.142292 
 0.024897 
 0.144356 
 
 
 7 YR 
 -0.176420 
 0.032728 
 0.179294 
 
 
 10 YR 
 -0.078536 
 0.040235 
 0.087825 
 
 
 20 YR 
 -0.023884 
 0.048389 
 0.052967 
 
 
 30 YR 
 0.103988 
 0.049662 
 0.114750 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122433 
 0.011042 
 0.122907 
 
 
 3 MO 
 -0.050238 
 0.005486 
 0.050523 
 
 
 6 MO 
 0.045311 
 0.005290 
 0.045604 
 
 
 1 YR 
 0.128469 
 0.007927 
 0.128703 
 
 
 2 YR 
 0.147937 
 0.011846 
 0.148389 
 
 
 3 YR 
 0.036255 
 0.014649 
 0.038978 
 
 
 5 YR 
 -0.127672 
 0.026852 
 0.130339 
 
 
 7 YR 
 -0.157564 
 0.034702 
 0.161170 
 
 
 10 YR 
 -0.055568 
 0.042502 
 0.069370 
 
 
 20 YR 
 0.004671 
 0.052066 
 0.051083 
 
 
 30 YR 
 0.134468 
 0.054220 
 0.144526

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.117042 
 0.016479 
 0.118142 
 
 
 3 MO 
 -0.042199 
 0.007148 
 0.042772 
 
 
 6 MO 
 0.048006 
 0.010840 
 0.049158 
 
 
 1 YR 
 0.119568 
 0.014281 
 0.120377 
 
 
 2 YR 
 0.132380 
 0.023073 
 0.134281 
 
 
 3 YR 
 0.033320 
 0.024763 
 0.041161 
 
 
 5 YR 
 -0.086719 
 0.042221 
 0.096010 
 
 
 7 YR 
 -0.107040 
 0.051757 
 0.118359 
 
 
 10 YR 
 -0.014808 
 0.052684 
 0.053504 
 
 
 20 YR 
 0.028806 
 0.056592 
 0.062289 
 
 
 30 YR 
 0.148055 
 0.060172 
 0.159275 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.115521 
 0.018392 
 0.116907 
 
 
 3 MO 
 -0.041013 
 0.008100 
 0.041768 
 
 
 6 MO 
 0.048737 
 0.010040 
 0.049712 
 
 
 1 YR 
 0.119534 
 0.013257 
 0.120232 
 
 
 2 YR 
 0.131260 
 0.022811 
 0.133134 
 
 
 3 YR 
 0.031511 
 0.024675 
 0.039659 
 
 
 5 YR 
 -0.089264 
 0.041794 
 0.098141 
 
 
 7 YR 
 -0.109917 
 0.051227 
 0.120751 
 
 
 10 YR 
 -0.017903 
 0.052573 
 0.054340 
 
 
 20 YR 
 0.025505 
 0.057677 
 0.061796 
 
 
 30 YR 
 0.144690 
 0.062075 
 0.156860

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.119776 
 0.012005 
 0.120348 
 
 
 3 MO 
 -0.050572 
 0.006454 
 0.050963 
 
 
 6 MO 
 0.045778 
 0.005203 
 0.046059 
 
 
 1 YR 
 0.126620 
 0.008537 
 0.126894 
 
 
 2 YR 
 0.141712 
 0.010750 
 0.142100 
 
 
 3 YR 
 0.023351 
 0.011988 
 0.026118 
 
 
 5 YR 
 -0.146240 
 0.020946 
 0.147662 
 
 
 7 YR 
 -0.177719 
 0.028646 
 0.179905 
 
 
 10 YR 
 -0.080515 
 0.033586 
 0.086931 
 
 
 20 YR 
 -0.026606 
 0.040891 
 0.047962 
 
 
 30 YR 
 0.097440 
 0.040913 
 0.105303 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.121681 
 0.012502 
 0.122291 
 
 
 3 MO 
 -0.051782 
 0.006800 
 0.052205 
 
 
 6 MO 
 0.045605 
 0.005466 
 0.045916 
 
 
 1 YR 
 0.128476 
 0.008467 
 0.128741 
 
 
 2 YR 
 0.147337 
 0.011031 
 0.147729 
 
 
 3 YR 
 0.032262 
 0.013334 
 0.034788 
 
 
 5 YR 
 -0.132207 
 0.022534 
 0.134023 
 
 
 7 YR 
 -0.160136 
 0.029658 
 0.162731 
 
 
 10 YR 
 -0.059528 
 0.035336 
 0.068795 
 
 
 20 YR 
 -0.001033 
 0.042491 
 0.041480 
 
 
 30 YR 
 0.124591 
 0.044272 
 0.131869

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.117787 
 0.009584 
 0.118157 
 
 
 3 MO 
 -0.049315 
 0.006058 
 0.049668 
 
 
 6 MO 
 0.046266 
 0.005298 
 0.046555 
 
 
 1 YR 
 0.127269 
 0.008371 
 0.127531 
 
 
 2 YR 
 0.141023 
 0.011379 
 0.141460 
 
 
 3 YR 
 0.023739 
 0.012087 
 0.026508 
 
 
 5 YR 
 -0.146612 
 0.020526 
 0.147974 
 
 
 7 YR 
 -0.178987 
 0.028140 
 0.181081 
 
 
 10 YR 
 -0.081587 
 0.033174 
 0.087776 
 
 
 20 YR 
 -0.027692 
 0.040004 
 0.047864 
 
 
 30 YR 
 0.097183 
 0.040506 
 0.104915 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.120546 
 0.010034 
 0.120943 
 
 
 3 MO 
 -0.051342 
 0.006315 
 0.051710 
 
 
 6 MO 
 0.045335 
 0.005631 
 0.045666 
 
 
 1 YR 
 0.128486 
 0.008473 
 0.128752 
 
 
 2 YR 
 0.146250 
 0.011558 
 0.146685 
 
 
 3 YR 
 0.032477 
 0.013354 
 0.034994 
 
 
 5 YR 
 -0.132386 
 0.022448 
 0.134186 
 
 
 7 YR 
 -0.160948 
 0.029467 
 0.163497 
 
 
 10 YR 
 -0.059888 
 0.035258 
 0.069068 
 
 
 20 YR 
 -0.001054 
 0.042480 
 0.041469 
 
 
 30 YR 
 0.125520 
 0.044828 
 0.132925

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.123233 
 0.010784 
 0.123681 
 
 
 3 MO 
 -0.044056 
 0.006324 
 0.044485 
 
 
 6 MO 
 0.050496 
 0.007299 
 0.050994 
 
 
 1 YR 
 0.124958 
 0.010597 
 0.125384 
 
 
 2 YR 
 0.138354 
 0.018575 
 0.139534 
 
 
 3 YR 
 0.032907 
 0.020750 
 0.038625 
 
 
 5 YR 
 -0.103471 
 0.033842 
 0.108601 
 
 
 7 YR 
 -0.128585 
 0.042620 
 0.135128 
 
 
 10 YR 
 -0.032748 
 0.047223 
 0.056489 
 
 
 20 YR 
 0.020092 
 0.051021 
 0.053635 
 
 
 30 YR 
 0.148636 
 0.053119 
 0.157395 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122749 
 0.012341 
 0.123337 
 
 
 3 MO 
 -0.043769 
 0.006824 
 0.044272 
 
 
 6 MO 
 0.050522 
 0.007259 
 0.051015 
 
 
 1 YR 
 0.124577 
 0.009247 
 0.124903 
 
 
 2 YR 
 0.137484 
 0.018146 
 0.138617 
 
 
 3 YR 
 0.031814 
 0.020700 
 0.037673 
 
 
 5 YR 
 -0.104640 
 0.033811 
 0.109706 
 
 
 7 YR 
 -0.129664 
 0.042356 
 0.136078 
 
 
 10 YR 
 -0.033663 
 0.047186 
 0.056994 
 
 
 20 YR 
 0.019469 
 0.052121 
 0.054405 
 
 
 30 YR 
 0.148119 
 0.054790 
 0.157452

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.117829 
 0.015936 
 0.118853 
 
 
 3 MO 
 -0.042341 
 0.007015 
 0.042892 
 
 
 6 MO 
 0.047561 
 0.010645 
 0.048685 
 
 
 1 YR 
 0.119692 
 0.014075 
 0.120479 
 
 
 2 YR 
 0.132157 
 0.022246 
 0.133932 
 
 
 3 YR 
 0.032291 
 0.024067 
 0.039945 
 
 
 5 YR 
 -0.090251 
 0.043918 
 0.099932 
 
 
 7 YR 
 -0.111606 
 0.054344 
 0.123592 
 
 
 10 YR 
 -0.019409 
 0.055020 
 0.057152 
 
 
 20 YR 
 0.024909 
 0.057070 
 0.061069 
 
 
 30 YR 
 0.145014 
 0.059178 
 0.156115 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.116083 
 0.018142 
 0.117429 
 
 
 3 MO 
 -0.040930 
 0.007914 
 0.041654 
 
 
 6 MO 
 0.048520 
 0.009851 
 0.049465 
 
 
 1 YR 
 0.119901 
 0.013052 
 0.120578 
 
 
 2 YR 
 0.131332 
 0.022264 
 0.133121 
 
 
 3 YR 
 0.030839 
 0.024286 
 0.038911 
 
 
 5 YR 
 -0.092318 
 0.043230 
 0.101521 
 
 
 7 YR 
 -0.113908 
 0.053382 
 0.125280 
 
 
 10 YR 
 -0.021830 
 0.054513 
 0.057560 
 
 
 20 YR 
 0.022421 
 0.058116 
 0.061046 
 
 
 30 YR 
 0.142511 
 0.061436 
 0.154635

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.119761 
 0.013871 
 0.120525 
 
 
 3 MO 
 -0.042822 
 0.006816 
 0.043336 
 
 
 6 MO 
 0.048926 
 0.010133 
 0.049918 
 
 
 1 YR 
 0.121234 
 0.013824 
 0.121984 
 
 
 2 YR 
 0.135256 
 0.021353 
 0.136856 
 
 
 3 YR 
 0.032721 
 0.024161 
 0.040347 
 
 
 5 YR 
 -0.093937 
 0.042818 
 0.102831 
 
 
 7 YR 
 -0.116356 
 0.053269 
 0.127465 
 
 
 10 YR 
 -0.023250 
 0.055091 
 0.058631 
 
 
 20 YR 
 0.023328 
 0.057144 
 0.060508 
 
 
 30 YR 
 0.146012 
 0.058786 
 0.156902 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.118392 
 0.015990 
 0.119419 
 
 
 3 MO 
 -0.041765 
 0.007623 
 0.042424 
 
 
 6 MO 
 0.049565 
 0.009360 
 0.050402 
 
 
 1 YR 
 0.121193 
 0.012700 
 0.121827 
 
 
 2 YR 
 0.134319 
 0.021187 
 0.135904 
 
 
 3 YR 
 0.031282 
 0.024411 
 0.039336 
 
 
 5 YR 
 -0.095789 
 0.042305 
 0.104326 
 
 
 7 YR 
 -0.118300 
 0.052456 
 0.128924 
 
 
 10 YR 
 -0.025174 
 0.054610 
 0.058995 
 
 
 20 YR 
 0.021527 
 0.058185 
 0.060787 
 
 
 30 YR 
 0.144262 
 0.060940 
 0.156065

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.117673 
 0.010589 
 0.118126 
 
 
 3 MO 
 -0.044837 
 0.005629 
 0.045173 
 
 
 6 MO 
 0.050650 
 0.005579 
 0.050942 
 
 
 1 YR 
 0.132682 
 0.007987 
 0.132911 
 
 
 2 YR 
 0.151230 
 0.011001 
 0.151611 
 
 
 3 YR 
 0.039667 
 0.014815 
 0.042220 
 
 
 5 YR 
 -0.123084 
 0.024640 
 0.125411 
 
 
 7 YR 
 -0.154431 
 0.032711 
 0.157696 
 
 
 10 YR 
 -0.053868 
 0.039653 
 0.066327 
 
 
 20 YR 
 0.004902 
 0.046724 
 0.045861 
 
 
 30 YR 
 0.134593 
 0.045671 
 0.141781 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122746 
 0.010673 
 0.123187 
 
 
 3 MO 
 -0.050042 
 0.005432 
 0.050322 
 
 
 6 MO 
 0.045344 
 0.005439 
 0.045653 
 
 
 1 YR 
 0.127448 
 0.007383 
 0.127652 
 
 
 2 YR 
 0.146850 
 0.010673 
 0.147219 
 
 
 3 YR 
 0.036583 
 0.014482 
 0.039218 
 
 
 5 YR 
 -0.123420 
 0.024787 
 0.125769 
 
 
 7 YR 
 -0.152480 
 0.033007 
 0.155846 
 
 
 10 YR 
 -0.049512 
 0.040059 
 0.063086 
 
 
 20 YR 
 0.012688 
 0.047589 
 0.048144 
 
 
 30 YR 
 0.143574 
 0.046847 
 0.150677

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122302 
 0.010376 
 0.122722 
 
 
 3 MO 
 -0.044528 
 0.006566 
 0.044989 
 
 
 6 MO 
 0.050711 
 0.006246 
 0.051078 
 
 
 1 YR 
 0.126543 
 0.009357 
 0.126874 
 
 
 2 YR 
 0.140070 
 0.016127 
 0.140955 
 
 
 3 YR 
 0.034242 
 0.018640 
 0.038792 
 
 
 5 YR 
 -0.109267 
 0.033940 
 0.114198 
 
 
 7 YR 
 -0.136155 
 0.043032 
 0.142511 
 
 
 10 YR 
 -0.038932 
 0.046548 
 0.059902 
 
 
 20 YR 
 0.016202 
 0.049427 
 0.050984 
 
 
 30 YR 
 0.147394 
 0.049157 
 0.155037 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122628 
 0.011296 
 0.123124 
 
 
 3 MO 
 -0.044961 
 0.007451 
 0.045548 
 
 
 6 MO 
 0.050151 
 0.007032 
 0.050620 
 
 
 1 YR 
 0.125824 
 0.008024 
 0.126069 
 
 
 2 YR 
 0.139300 
 0.015779 
 0.140152 
 
 
 3 YR 
 0.033612 
 0.018695 
 0.038264 
 
 
 5 YR 
 -0.109432 
 0.033823 
 0.114322 
 
 
 7 YR 
 -0.135868 
 0.042561 
 0.142101 
 
 
 10 YR 
 -0.038141 
 0.046201 
 0.059131 
 
 
 20 YR 
 0.017737 
 0.050026 
 0.052042 
 
 
 30 YR 
 0.149189 
 0.050409 
 0.157124

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.121337 
 0.010258 
 0.121751 
 
 
 3 MO 
 -0.046780 
 0.005138 
 0.047049 
 
 
 6 MO 
 0.048984 
 0.005606 
 0.049290 
 
 
 1 YR 
 0.129005 
 0.007477 
 0.129212 
 
 
 2 YR 
 0.146845 
 0.012332 
 0.147340 
 
 
 3 YR 
 0.036443 
 0.014780 
 0.039205 
 
 
 5 YR 
 -0.123390 
 0.027942 
 0.126380 
 
 
 7 YR 
 -0.155943 
 0.036000 
 0.159868 
 
 
 10 YR 
 -0.058004 
 0.041406 
 0.070742 
 
 
 20 YR 
 -0.001465 
 0.046974 
 0.045965 
 
 
 30 YR 
 0.129451 
 0.047476 
 0.137526 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.123188 
 0.010123 
 0.123585 
 
 
 3 MO 
 -0.048770 
 0.005694 
 0.049087 
 
 
 6 MO 
 0.046899 
 0.006427 
 0.047318 
 
 
 1 YR 
 0.127057 
 0.007187 
 0.127251 
 
 
 2 YR 
 0.146004 
 0.012299 
 0.146498 
 
 
 3 YR 
 0.037221 
 0.014847 
 0.039953 
 
 
 5 YR 
 -0.119227 
 0.028145 
 0.122363 
 
 
 7 YR 
 -0.148978 
 0.036073 
 0.153099 
 
 
 10 YR 
 -0.048104 
 0.041512 
 0.062947 
 
 
 20 YR 
 0.012617 
 0.047674 
 0.048303 
 
 
 30 YR 
 0.144990 
 0.048451 
 0.152537

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.124092 
 0.009497 
 0.124439 
 
 
 3 MO 
 -0.045516 
 0.006087 
 0.045903 
 
 
 6 MO 
 0.050727 
 0.005399 
 0.051000 
 
 
 1 YR 
 0.128710 
 0.006766 
 0.128880 
 
 
 2 YR 
 0.143438 
 0.015440 
 0.144229 
 
 
 3 YR 
 0.037329 
 0.017416 
 0.041025 
 
 
 5 YR 
 -0.115494 
 0.030593 
 0.119299 
 
 
 7 YR 
 -0.145121 
 0.038787 
 0.149987 
 
 
 10 YR 
 -0.047566 
 0.043202 
 0.063593 
 
 
 20 YR 
 0.011574 
 0.047391 
 0.047726 
 
 
 30 YR 
 0.145484 
 0.046883 
 0.152524 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.125126 
 0.009217 
 0.125450 
 
 
 3 MO 
 -0.046730 
 0.006980 
 0.047225 
 
 
 6 MO 
 0.049314 
 0.006595 
 0.049733 
 
 
 1 YR 
 0.127109 
 0.006267 
 0.127257 
 
 
 2 YR 
 0.142019 
 0.015207 
 0.142794 
 
 
 3 YR 
 0.036477 
 0.017587 
 0.040321 
 
 
 5 YR 
 -0.114926 
 0.030739 
 0.118785 
 
 
 7 YR 
 -0.143286 
 0.038574 
 0.148159 
 
 
 10 YR 
 -0.044359 
 0.043070 
 0.061143 
 
 
 20 YR 
 0.016770 
 0.047785 
 0.049606 
 
 
 30 YR 
 0.151375 
 0.047387 
 0.158297

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.115186 
 0.013369 
 0.115923 
 
 
 3 MO 
 -0.047029 
 0.010570 
 0.048147 
 
 
 6 MO 
 0.045597 
 0.005951 
 0.045965 
 
 
 1 YR 
 0.127906 
 0.009997 
 0.128278 
 
 
 2 YR 
 0.132743 
 0.014193 
 0.133463 
 
 
 3 YR 
 0.021324 
 0.011729 
 0.024202 
 
 
 5 YR 
 -0.122341 
 0.024944 
 0.124739 
 
 
 7 YR 
 -0.128714 
 0.038720 
 0.134145 
 
 
 10 YR 
 -0.030756 
 0.033940 
 0.045199 
 
 
 20 YR 
 0.016125 
 0.031796 
 0.034970 
 
 
 30 YR 
 0.121798 
 0.032242 
 0.125797 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.115894 
 0.014070 
 0.116704 
 
 
 3 MO 
 -0.047533 
 0.010582 
 0.048641 
 
 
 6 MO 
 0.045336 
 0.006374 
 0.045760 
 
 
 1 YR 
 0.127940 
 0.010765 
 0.128371 
 
 
 2 YR 
 0.132855 
 0.016091 
 0.133779 
 
 
 3 YR 
 0.021141 
 0.013638 
 0.024982 
 
 
 5 YR 
 -0.123460 
 0.024540 
 0.125761 
 
 
 7 YR 
 -0.130733 
 0.037271 
 0.135699 
 
 
 10 YR 
 -0.033778 
 0.033156 
 0.046775 
 
 
 20 YR 
 0.011628 
 0.034520 
 0.035639 
 
 
 30 YR 
 0.116783 
 0.037076 
 0.122260

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.123103 
 0.010555 
 0.123534 
 
 
 3 MO 
 -0.045060 
 0.006606 
 0.045520 
 
 
 6 MO 
 0.049568 
 0.007690 
 0.050134 
 
 
 1 YR 
 0.124890 
 0.009951 
 0.125268 
 
 
 2 YR 
 0.138267 
 0.017571 
 0.139329 
 
 
 3 YR 
 0.032326 
 0.020236 
 0.037893 
 
 
 5 YR 
 -0.107007 
 0.034205 
 0.112104 
 
 
 7 YR 
 -0.132952 
 0.043494 
 0.139578 
 
 
 10 YR 
 -0.037095 
 0.047215 
 0.059194 
 
 
 20 YR 
 0.015873 
 0.050610 
 0.051932 
 
 
 30 YR 
 0.145748 
 0.051416 
 0.154162 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122606 
 0.011913 
 0.123157 
 
 
 3 MO 
 -0.044748 
 0.007423 
 0.045332 
 
 
 6 MO 
 0.049644 
 0.008009 
 0.050257 
 
 
 1 YR 
 0.124616 
 0.008827 
 0.124914 
 
 
 2 YR 
 0.137644 
 0.017359 
 0.138685 
 
 
 3 YR 
 0.031629 
 0.020457 
 0.037415 
 
 
 5 YR 
 -0.107515 
 0.034072 
 0.112551 
 
 
 7 YR 
 -0.133170 
 0.043002 
 0.139640 
 
 
 10 YR 
 -0.036946 
 0.046933 
 0.058886 
 
 
 20 YR 
 0.016595 
 0.051555 
 0.053034 
 
 
 30 YR 
 0.146674 
 0.053137 
 0.155591

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122686 
 0.009646 
 0.123046 
 
 
 3 MO 
 -0.044193 
 0.005745 
 0.044547 
 
 
 6 MO 
 0.051630 
 0.005293 
 0.051888 
 
 
 1 YR 
 0.129238 
 0.007001 
 0.129418 
 
 
 2 YR 
 0.142980 
 0.015852 
 0.143814 
 
 
 3 YR 
 0.037006 
 0.017944 
 0.040940 
 
 
 5 YR 
 -0.114923 
 0.031324 
 0.118919 
 
 
 7 YR 
 -0.143620 
 0.039428 
 0.148685 
 
 
 10 YR 
 -0.045450 
 0.043896 
 0.062457 
 
 
 20 YR 
 0.014642 
 0.048139 
 0.049207 
 
 
 30 YR 
 0.148996 
 0.047474 
 0.156033 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.124591 
 0.009088 
 0.124907 
 
 
 3 MO 
 -0.046099 
 0.006477 
 0.046530 
 
 
 6 MO 
 0.049752 
 0.006421 
 0.050145 
 
 
 1 YR 
 0.127494 
 0.006150 
 0.127635 
 
 
 2 YR 
 0.141698 
 0.015506 
 0.142503 
 
 
 3 YR 
 0.036289 
 0.017998 
 0.040316 
 
 
 5 YR 
 -0.114545 
 0.031444 
 0.118584 
 
 
 7 YR 
 -0.142369 
 0.039280 
 0.147439 
 
 
 10 YR 
 -0.043301 
 0.043840 
 0.060872 
 
 
 20 YR 
 0.018058 
 0.048571 
 0.050724 
 
 
 30 YR 
 0.152852 
 0.048035 
 0.159879

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.115581 
 0.013995 
 0.116383 
 
 
 3 MO 
 -0.046435 
 0.010861 
 0.047626 
 
 
 6 MO 
 0.044736 
 0.005540 
 0.045060 
 
 
 1 YR 
 0.127748 
 0.010448 
 0.128154 
 
 
 2 YR 
 0.132417 
 0.013896 
 0.133108 
 
 
 3 YR 
 0.020244 
 0.012420 
 0.023587 
 
 
 5 YR 
 -0.123516 
 0.026284 
 0.126144 
 
 
 7 YR 
 -0.129121 
 0.039968 
 0.134869 
 
 
 10 YR 
 -0.030873 
 0.035066 
 0.046057 
 
 
 20 YR 
 0.013453 
 0.033600 
 0.035405 
 
 
 30 YR 
 0.118719 
 0.033625 
 0.123160 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.116425 
 0.015197 
 0.117363 
 
 
 3 MO 
 -0.047130 
 0.010648 
 0.048259 
 
 
 6 MO 
 0.044212 
 0.005728 
 0.044563 
 
 
 1 YR 
 0.127418 
 0.010767 
 0.127850 
 
 
 2 YR 
 0.132069 
 0.015634 
 0.132945 
 
 
 3 YR 
 0.019590 
 0.013913 
 0.023825 
 
 
 5 YR 
 -0.125030 
 0.025543 
 0.127484 
 
 
 7 YR 
 -0.131432 
 0.038386 
 0.136654 
 
 
 10 YR 
 -0.034061 
 0.034297 
 0.047724 
 
 
 20 YR 
 0.008985 
 0.035641 
 0.035882 
 
 
 30 YR 
 0.113803 
 0.037775 
 0.119611

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122451 
 0.010581 
 0.122888 
 
 
 3 MO 
 -0.045261 
 0.005896 
 0.045627 
 
 
 6 MO 
 0.050894 
 0.005302 
 0.051158 
 
 
 1 YR 
 0.130394 
 0.007569 
 0.130604 
 
 
 2 YR 
 0.145544 
 0.015932 
 0.146376 
 
 
 3 YR 
 0.038152 
 0.016594 
 0.041460 
 
 
 5 YR 
 -0.116628 
 0.028074 
 0.119817 
 
 
 7 YR 
 -0.147888 
 0.035616 
 0.151935 
 
 
 10 YR 
 -0.050877 
 0.039794 
 0.064056 
 
 
 20 YR 
 0.007252 
 0.044059 
 0.043697 
 
 
 30 YR 
 0.140284 
 0.043575 
 0.146615 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.124278 
 0.009939 
 0.124658 
 
 
 3 MO 
 -0.047406 
 0.006639 
 0.047848 
 
 
 6 MO 
 0.048384 
 0.006467 
 0.048796 
 
 
 1 YR 
 0.127479 
 0.006555 
 0.127640 
 
 
 2 YR 
 0.142693 
 0.015322 
 0.143478 
 
 
 3 YR 
 0.035989 
 0.016561 
 0.039466 
 
 
 5 YR 
 -0.116894 
 0.028369 
 0.120142 
 
 
 7 YR 
 -0.146401 
 0.035657 
 0.150497 
 
 
 10 YR 
 -0.047469 
 0.040018 
 0.061523 
 
 
 20 YR 
 0.013466 
 0.044960 
 0.045987 
 
 
 30 YR 
 0.147481 
 0.044624 
 0.153803

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.122212 
 0.010821 
 0.122669 
 
 
 3 MO 
 -0.044718 
 0.006534 
 0.045172 
 
 
 6 MO 
 0.049650 
 0.007534 
 0.050194 
 
 
 1 YR 
 0.125434 
 0.010253 
 0.125835 
 
 
 2 YR 
 0.138402 
 0.017231 
 0.139424 
 
 
 3 YR 
 0.032129 
 0.019711 
 0.037469 
 
 
 5 YR 
 -0.108617 
 0.033980 
 0.113587 
 
 
 7 YR 
 -0.134983 
 0.043249 
 0.141455 
 
 
 10 YR 
 -0.038940 
 0.046621 
 0.059961 
 
 
 20 YR 
 0.013826 
 0.049844 
 0.050671 
 
 
 30 YR 
 0.144050 
 0.050326 
 0.152227 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.121949 
 0.012059 
 0.122517 
 
 
 3 MO 
 -0.044606 
 0.007284 
 0.045172 
 
 
 6 MO 
 0.049572 
 0.007833 
 0.050161 
 
 
 1 YR 
 0.125097 
 0.008927 
 0.125402 
 
 
 2 YR 
 0.137873 
 0.016995 
 0.138871 
 
 
 3 YR 
 0.031656 
 0.019987 
 0.037205 
 
 
 5 YR 
 -0.108708 
 0.033777 
 0.113616 
 
 
 7 YR 
 -0.134655 
 0.042612 
 0.140957 
 
 
 10 YR 
 -0.038124 
 0.046200 
 0.059120 
 
 
 20 YR 
 0.015376 
 0.050709 
 0.051923 
 
 
 30 YR 
 0.145858 
 0.052063 
 0.154491

mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.090491 
 0.018533 
 0.092281 
 
 
 3 MO 
 -0.037779 
 0.007696 
 0.038518 
 
 
 6 MO 
 0.034054 
 0.012407 
 0.036142 
 
 
 1 YR 
 0.095662 
 0.019014 
 0.097446 
 
 
 2 YR 
 0.100806 
 0.030550 
 0.105122 
 
 
 3 YR 
 0.015970 
 0.028418 
 0.032002 
 
 
 5 YR 
 -0.056250 
 0.042464 
 0.069867 
 
 
 7 YR 
 -0.072999 
 0.047974 
 0.086722 
 
 
 10 YR 
 -0.002512 
 0.048529 
 0.047426 
 
 
 20 YR 
 0.027260 
 0.053648 
 0.059027 
 
 
 30 YR 
 0.114475 
 0.059843 
 0.128511 
 
 
 
 
 
 
 mean 
 std 
 RMSE 
 
 
 
 
 1 MO 
 -0.084302 
 0.020397 
 0.086620 
 
 
 3 MO 
 -0.032739 
 0.008809 
 0.033849 
 
 
 6 MO 
 0.037524 
 0.012025 
 0.039316 
 
 
 1 YR 
 0.096474 
 0.019692 
 0.098370 
 
 
 2 YR 
 0.097769 
 0.032711 
 0.102849 
 
 
 3 YR 
 0.010418 
 0.030273 
 0.031326 
 
 
 5 YR 
 -0.064620 
 0.041785 
 0.076411 
 
 
 7 YR 
 -0.082738 
 0.047705 
 0.094937 
 
 
 10 YR 
 -0.013243 
 0.050073 
 0.050629 
 
 
 20 YR 
 0.015481 
 0.055901 
 0.056708 
 
 
 30 YR 
 0.102359 
 0.063831 
 0.119824

* From the comparsion tables above, **generally speaking, the AR(1) model has more out-of-sample accuracy than Random Walk model**. But the real accuracy **depends on specific dates and yield maturities**.
* **Solution to Step 3**: Generally speaking, these forecasts perform better than those with generated under a Random Walk.